contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,380 @@
+
+
+## Design Details
+
+### Snapshot API Design
+
+The API design of VolumeSnapshot and VolumeSnapshotData is modeled after PersistentVolumeClaim and PersistentVolume.
+
+#### The `VolumeSnapshot` Object
+
+```GO
+
+// The volume snapshot object accessible to the user. Upon successful creation of the actual
+// snapshot by the volume provider it is bound to the corresponding VolumeSnapshotData through
+// the VolumeSnapshotSpec
+type VolumeSnapshot struct {
+	metav1.TypeMeta `json:",inline"`
+	Metadata        metav1.ObjectMeta `json:"metadata"`
+
+	// Spec represents the desired state of the snapshot
+	// +optional
+	Spec VolumeSnapshotSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status represents the latest observer state of the snapshot
+	// +optional
+	Status VolumeSnapshotStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+type VolumeSnapshotList struct {
+	metav1.TypeMeta `json:",inline"`
+	Metadata        metav1.ListMeta  `json:"metadata"`
+	Items           []VolumeSnapshot `json:"items"`
+}
+
+// VolumeSnapshotSpec is the desired state of the volume snapshot
+type VolumeSnapshotSpec struct {
+        // PersistentVolumeClaimName is the name of the PVC being snapshotted
+        // +optional
+        PersistentVolumeClaimName string `json:"persistentVolumeClaimName" protobuf:"bytes,1,opt,name=persistentVolumeClaimName"`
+
+        // SnapshotDataName binds the VolumeSnapshot object with the VolumeSnapshotData
+        // +optional
+        SnapshotDataName string `json:"snapshotDataName" protobuf:"bytes,2,opt,name=snapshotDataName"`
+
+        // Name of the SnapshotClass required by the volume snapshot.
+        // +optional
+        SnapshotClassName string `json:"storageClassName" protobuf:"bytes,3,opt,name=storageClassName"`
+}
+
+type VolumeSnapshotStatus struct {
+	// The time the snapshot was successfully created
+	// +optional
+	CreationTimestamp metav1.Time `json:"creationTimestamp" protobuf:"bytes,1,opt,name=creationTimestamp"`
+
+	// Represents the lates available observations about the volume snapshot
+	Conditions []VolumeSnapshotCondition `json:"conditions" protobuf:"bytes,2,rep,name=conditions"`
+}
+
+type VolumeSnapshotConditionType string
+
+// These are valid conditions of a volume snapshot.
+const (
+        // VolumeSnapshotConditionCreating means the snapshot is being created but
+        // it is not cut yet.
+        VolumeSnapshotConditionCreating VolumeSnapshotConditionType = "Creating"
```

Comment:
API needs to return immediately while CSI driver call is blocking until the snapshot is cut.  That's why we need a "Creating" status for API.  It might map to UNKNOWN in CSI.